### PR TITLE
Added Hinge Loss to loss functions

### PR DESCRIPTION
**Pull request by Abilityguy**
Reasons for this PR:
1. The Hinge Loss function is currently implemented as ```Hinge Embedding Loss``` in mlpack. This implementation is only correct for 1D tensors and doesn't give the right answers for tensors that are higher than 1 dimension. This [colab notebook](https://colab.research.google.com/drive/1iA1pHNxRE9fVxB01oJUEiM2BTCACQHmK?usp=sharing) highlights this by comparing with the [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/losses/hinge) implementation. Also, the current implementation only incorporates mean reduction. This PR adds an additional ```reduction``` parameter to choose between mean and sum reduction.

2. Issue #2444 and PR #2495 aim to change the implementation of ```Hinge Embedding Loss``` to the Hinge Embedding Loss function in [pytorch](https://pytorch.org/docs/stable/generated/torch.nn.HingeEmbeddingLoss.html). Hence, a new implementation file with the corrections implemented has been made for the Hinge Loss function.

Let me know if you have any suggestions or changes on this.

### CODE REVIEW

**Comment by zoq** [comment]
Thanks for working on this, are you planning to implement a test case for the loss function as well?

**Comment by Abilityguy** [comment]
Yes, will look into adding tests.

**Comment by Abilityguy** [comment]
Have added a few tests. 
Let me know if you have any suggestions or changes on this.

File: src/mlpack/methods/ann/loss_functions/hinge_loss.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+#include<mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Computes the hinge loss between y_true and y_pred. Expects y_true to be
+ * either -1 or 1. If y_true is either 0 or 1, a temporary conversion is made to
+ * calculate the loss.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_loss.hpp:24]: Do you think you can use doxygen math `$$` around `y_true` and `y_pred` also maybe it would be useful to write the complete formula here. Could make it easier to understand what `y_true` and `y_pred` is and also how they fit into everything.

File: src/mlpack/methods/ann/loss_functions/hinge_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+template <
+        typename InputDataType = arma::mat,
+        typename OutputDataType = arma::mat
+>
+class HingeLoss
+{
+ public:
+  /**
+   * Create HingeLoss object.
+   * @param reduction Specifies the reduction to apply to the output. If false,
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_loss.hpp:40]:
```suggestion
   * Create HingeLoss object.
   *
   * @param reduction Specifies the reduction to apply to the output. If false,
```

Just a single empty line to split the method description from the parameter description.

File: src/mlpack/methods/ann/loss_functions/hinge_loss.hpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  /**
+   * Computes the Hinge loss function.
+   *
+   * @param prediction Prediction used for evaluating the specified loss
+   *     function.
+   * @param target Target data to compare with.
+   */
+  template<typename PredictionType, typename TargetType>
+  typename PredictionType::elem_type Forward(const PredictionType& prediction,
+                                              const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_loss.hpp:57]:
```suggestion
                                             const TargetType& target);
```

Minor style issue, remove an extra space.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 4]
```diff
@@ -955,3 +956,76 @@ TEST_CASE("TripletMarginLossTest")
...
+
+  input = {{0.90599973, -0.33040298, 0.07123354},
+      {0.71988434, 0.49657596, 0.39873373},
+      {-0.57646927, 0.3951491 , -0.1003365},
+      {0.12528634, 0.68122971, 0.85448826}};
+
+  target = {{-1, -1, 1},
+      {-1, 1, 1},
+      {1, -1, -1},
+      {1, -1, -1}};
```
Comment (zoq) [inline src/mlpack/tests/loss_functions_test.cpp:995]: Are those values calculated manually if yes or if they are calculated with some other tool, let's add a simple comment here.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 5]
```diff
@@ -955,3 +956,76 @@ TEST_CASE("TripletMarginLossTest")
...
+
+  // Binary labels for target.
+  target_b = {{0, 0, 1},
```
Comment (zoq) [inline src/mlpack/tests/loss_functions_test.cpp:998]: See comment above.

File: src/mlpack/methods/ann/loss_functions/hinge_loss.hpp (hunk 1 above)
Comment (Abilityguy) [reply src/mlpack/methods/ann/loss_functions/hinge_loss.hpp:24]: That's a great idea. I know about using ```\f$``` to type formulas in doxygen. Is this the same as ```$$```?
If not, could you point me to an example of this in the mlpack codebase?

Comment (zoq) [reply src/mlpack/methods/ann/loss_functions/hinge_loss.hpp]: Right, either is fine you can just go with`\f$`.

**Comment by Abilityguy** [comment]
Have added doxygen math ```\f$``` to y_true and y_pred. Also added the formula for Hinge Loss using the same.
Do let me know if there are any issues with the same.

**Review by zoq — APPROVED** [review]
No further comments from my side, thanks for putting this together.

**Review by shrit — APPROVED** [review]

**Comment by Abilityguy** [comment]
I believe this PR can be merged now.
Thanks for the quick reviews!
